awx_collection/plugins/modules/tower_token.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/python
+# coding: utf-8 -*-
+
+
+# (c) 2020, John Westcott IV <john.westcott.iv@redhat.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {'metadata_version': '1.1',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+DOCUMENTATION = '''
+---
+module: tower_token
+author: "John Westcott IV (@john-westcott-iv)"
+version_added: "2.3"
+short_description: create, update, or destroy Ansible Tower tokens.
+description:
+    - Create or destroy Ansible Tower tokens. See
+      U(https://www.ansible.com/tower) for an overview.
+    - If you create a token it is your responsibility to delete the token.
+options:
+    description:
+      description:
+        - Optional description of this access token.
+      required: False
+      type: str
+      default: ''
+    application:
+      description:
+        - The application tied to this token.
+      required: False
+      type: str
+    scope:
+      description:
+        - Allowed scopes, further restricts user's permissions. Must be a simple space-separated string with allowed scopes ['read', 'write'].
+      required: False
+      type: str
+      default: 'write'
+      choices: ["read", "write"]
+    existing_token:
+      description: An existing token (for use with state absent)
+      type: dict
+    state:
+      description:
+        - Desired state of the resource.
+      choices: ["present", "absent"]
+      default: "present"
+      type: str
+extends_documentation_fragment: awx.awx.auth
+'''
+
+EXAMPLES = '''
+- block:
+    - name: Create a new token using an existing token
+      tower_token:
+        description: '{{ token_description }}'
+        scope: "write"
+        state: present
+        tower_oauthtoken: "{{ ny_existing_token }}"
+
+    - name: Delete this token
+      tower_token:
+        existing_token: "{{ tower_token }}"
+        state: absent
+
+    - name: Create a new token using username/password
+      tower_token:
+        description: '{{ token_description }}'
+        scope: "write"
+        state: present
+        tower_username: "{{ ny_username }}"
+        tower_password: "{{ ny_password }}"
+
+    - name: Use our new token to make another call
+      tower_job_list:
+        tower_oauthtoken: "{{ tower_token }}"
+
+  always:
+    - name: Delete our Token with the token we created
+      tower_token:
+        existing_token: "{{ tower_token }}"
+        state: absent
+      when: tower_token is defined
+'''
+
+RETURNS = '''
+tower_token:
+  type: dict
+  description: A Tower token object which can be used for auth or token deletion
```

Comment:
The ovirt example you link documents the type as `complex` and gives the sub-keys

https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L158

I actually think that would be good. Even if it gave more keys than what are documented, this would be a good way to specify the 2 functional items - the token and the token_id. In the parameter documentation we discussed

> A dictionary structure as returned by the tower_token module.

It just defers here. We also don't have any real examples of the dictionary typed out. I think it needs to be somewhere more clearly stated that (1) the function of this data structure is to contain the token and the token id, and what the key names for those are and (2) also mention in the return docs that this is returned both in this key and in the ansible_facts.

Reply:
I'm a little torn on providing additional details about the return dict. The whole point of the module is to hide the login mechanics from someone. IMHO we should be trying to document "here is how you call this thing and pass it onto the next" vs. "here is how you call this thing, here are all of the internals of what you get back, now pass it onto the next". If there was other useful data within the dick maybe but I'm not sure the fields would provide any use. And people really should not have a need to "create" one of these dicts on their own to use for authentication (that would likely be more work than just cutting a token directly in Tower and using the string form). Do you see an advantage to documenting these files two fields?

As for the return 'token' value, I think that might just be an artifact from before getting the ansible_fact returned. Any objection to removing it from the response?

Reply:
Technically this module is the CRUD interface to the tokens endpoint and no other module includes a RETURN statement. But I think this one should because it’s setting the ansible_fact for token usage which deviates form the norm. In addition, the token string is only available though this fact once and can't be obtained through the API on subsequent calls.

Because of that, I pushed a change which modified the module description to add:
```
    - In addition, the module sets an Ansible fact which can be passed into other
      tower_* modules as the parameter tower_oauthtoken. See examples for usage.
    - Because of the sensitive nature of tokens, the created token value is only available once
      through the Ansible fact. (See RETURN for details)
```

Then In the return we say:
```
tower_token:
  type: dict
  description: An Ansible Fact variable representing a Tower token object which can be used for auth in subsequent modules. See examples for usage.
  contains:
    token:
      description: The token that was generated. This token can never be accessed again, make sure this value is noted before it is lost.
      type: str
  returned: on successful create
```

However, I’m not sure we want to dive into the data structure (i.e. the ID). Technically, like all other modules, the ID is in the return. But people really shouldn't care what else is in the dict and they shouldn't be trying to create one on their own. 

I also removed the token in the response (that was just an artifact from before the ansible_fact was returned). 

Let me know what people think about these changes.

Reply:
ok, resolved

this was the most important things out of the various discussions here